src/pygeoprocessing/geoprocessing.py

```diff
@@ -535,6 +535,72 @@ def raster_calculator(
                 pass
 
 
+def raster_reduce(function, raster_path_band, initializer,
+                  largest_block=_LARGEST_ITERBLOCK):
+    """Cumulatively apply a reducing function to each block of a raster.
+
+    This effectively reduces the entire raster to a single value, but it works
+    by blocks to be memory-efficient.
+
+    The ``function`` signature should be ``function(aggregator, block)``, where
+    ``aggregator`` is the aggregated value so far, and ``block`` is a 2D numpy
+    array containing the data from the block to reduce next.
+
+    ``function`` is called once on each block. On the first ``function`` call,
+    ``aggregator`` is initialized with ``initializer``. The return value from
+    each ``function`` call is passed in as the ``aggregator`` argument to the
+    subsequent ``function`` call. When all blocks have been reduced, the return
+    value of the final ``function`` call is returned.
+
+    Example:
+        Calculate the sum of all values in a raster::
+
+            raster_reduce(lambda total, block: total + numpy.sum(block),
+                          (raster_path, 1), 0)
+
+        Calculate a histogram of all values in a raster::
+
+            def add_to_histogram(histogram, block):
+                return histogram + numpy.histogram(block, bins=10)[0]
+
+            raster_reduce(add_to_histogram, (raster_path, 1), numpy.zeros(10))
+
+        Calculate the sum of all values in a raster, excluding nodata::
+
+            nodata = pygeoprocessing.get_raster_info(raster_path)['nodata'][0]
+            def sum_excluding_nodata(total, block):
+                return total + numpy.sum(block[block != nodata])
+
+            raster_reduce(sum_excluding_nodata, (raster_path, 1), 0)
+
+    Args:
+        function (func): function to apply to each raster block
+        raster_path_band (tuple): (path, band) tuple of the raster to reduce
+        initializer (obj): value to initialize the aggregator for the
+            first function call
+
+    Returns:
+        aggregate value, the final value returned from ``function``
+    """
+    aggregator = initializer
+    last_time = time.time()
+    pixels_processed = 0
+    x_size, y_size = get_raster_info(raster_path_band[0])['raster_size']
+    n_pixels = x_size * y_size
+    for (_, block) in iterblocks(raster_path_band,
+                                 largest_block=largest_block):
+        aggregator = function(aggregator, block)
+        pixels_processed += block.shape[0] * block.shape[1]
```

Comment:
Such a minor thing, but slightly cheaper :)

```suggestion
        pixels_processed += block.size
```